service/matching/matching_engine.go

```diff
@@ -2877,3 +2900,117 @@ func largerBacklogAge(rootBacklogAge *durationpb.Duration, currentPartitionAge *
 	}
 	return currentPartitionAge
 }
+
+func (e *matchingEngineImpl) UpdateTaskQueueConfig(ctx context.Context, request *matchingservice.UpdateTaskQueueConfigRequest) (*matchingservice.UpdateTaskQueueConfigResponse, error) {
+	taskQueueFamily, err := tqid.NewTaskQueueFamily(request.NamespaceId, request.UpdateTaskqueueConfig.GetTaskQueueName())
+	if err != nil {
+		return nil, err
+	}
+
+	taskQueueType := request.UpdateTaskqueueConfig.TaskQueueType
+	tqm, _, err := e.getTaskQueuePartitionManager(ctx,
+		taskQueueFamily.TaskQueue(taskQueueType).RootPartition(),
+		true, loadCauseOtherWrite)
+
+	if err != nil {
+		return nil, err
+	}
+
+	updateOptions := UserDataUpdateOptions{Source: "UpdateTaskQueueConfig"}
+
+	e.logger.Debug("Updating TaskQueueConfig")
+	_, err = tqm.GetUserDataManager().UpdateUserData(ctx, updateOptions,
+		func(tqud *persistencespb.TaskQueueUserData) (*persistencespb.TaskQueueUserData, bool, error) {
+			data := common.CloneProto(tqud)
+			if data == nil {
+				data = &persistencespb.TaskQueueUserData{}
+			}
+			if data.PerType == nil {
+				data.PerType = make(map[int32]*persistencespb.TaskQueueTypeUserData)
+			}
+			tqType := int32(request.UpdateTaskqueueConfig.TaskQueueType)
+
+			if data.PerType[tqType] == nil {
+				data.PerType[tqType] = &persistencespb.TaskQueueTypeUserData{}
+			}
+
+			if data.PerType[tqType].TaskQueueConfig == nil {
+				data.PerType[tqType].TaskQueueConfig = &taskqueuepb.TaskQueueConfig{}
+			}
+
+			// Last-writer-wins: check if incoming update is newer than stored tqud.Clock
+			existingClock := data.Clock
+			if existingClock == nil {
+				existingClock = hlc.Zero(e.clusterMeta.GetClusterID())
+			}
+
+			now := hlc.Next(existingClock, e.timeSource)
+
+			utcTimeNow := e.timeSource.Now().UTC()
+			// Update relevant config fields
+			cfg := data.PerType[tqType].TaskQueueConfig
+			changed := false
+
+			// Queue Rate Limit
+			queueRateLimit := request.UpdateTaskqueueConfig.UpdateQueueRateLimit
+			if queueRateLimit != nil {
+				queueRateLimitConfig := &taskqueuepb.RateLimitConfig{
+					RateLimit: &taskqueuepb.RateLimit{
+						RequestsPerSecond: queueRateLimit.RateLimit.RequestsPerSecond,
+					},
+					Metadata: &taskqueuepb.ConfigMetadata{
+						Reason:     queueRateLimit.RateLimit.Reason, // assuming getters are defined
+						Source:     "updateTaskQueueConfig",         // or "system"/"worker" depending on your context
```

Comment:
source shouldn't be stored.. it's always "api". source is just for returning in Describe